pkg/mpc-examples/contracts/01-weight-changer/WeightChangerController.sol

```diff
@@ -0,0 +1,146 @@
+    uint256 private constant _MIN_WEIGHT = 1e16; // 1%
+    uint256 private constant _MAX_WEIGHT = 99e16; // 99%
+
+    IVault private immutable _vault;
+    bytes32 private immutable _poolId;
+
+    constructor(IVault vault) {
+        // Get poolId from the factory
+        bytes32 poolId = IManagedPool(ILastCreatedPoolFactory(msg.sender).getLastCreatedPool()).getPoolId();
+
+        // Verify that this is a real Vault and the pool is registered - this call will revert if not.
+        vault.getPool(poolId);
+
+        // Set the global tokens variables
+        (IERC20[] memory tokens, , ) = vault.getPoolTokens(poolId);
+        _setTokens(tokens);
+
+        _vault = vault;
+        _poolId = poolId;
+    }
+
+    function make5050() public {
+        uint256[] memory newWeights = new uint256[](2);
+        newWeights[0] = 50e16;
+        newWeights[1] = 50e16;
+        _updateWeights(block.timestamp, block.timestamp + _REWEIGHT_DURATION, getTokens(), newWeights);
+    }
+
+    function make8020() public {
+        uint256[] memory newWeights = new uint256[](2);
+        newWeights[0] = 80e16;
+        newWeights[1] = 20e16;
+        _updateWeights(block.timestamp, block.timestamp + _REWEIGHT_DURATION, getTokens(), newWeights);
+    }
+
+    function make9901() public {
+        uint256[] memory newWeights = new uint256[](2);
+        newWeights[0] = 99e16;
+        newWeights[1] = 1e16;
+        _updateWeights(block.timestamp, block.timestamp + _REWEIGHT_DURATION, getTokens(), newWeights);
+    }
+
+    // === Public Getters ===
+    function getMaximumWeight() public pure returns (uint256) {
+        return _MAX_WEIGHT;
+    }
+
+    function getMinimumWeight() public pure returns (uint256) {
+        return _MIN_WEIGHT;
+    }
+
+    function getCurrentWeights() public view returns (uint256[] memory) {
+        return _getPool().getNormalizedWeights();
+    }
+
+    function getReweightDuration() public pure returns (uint256) {
+        return _REWEIGHT_DURATION;
+    }
+
+    function getTokens() public view returns (IERC20[] memory) {
+        return _tokens;
+    }
+
+    function getPoolId() public view returns (bytes32) {
+        return _poolId;
+    }
+
+    function getVault() public view returns (IVault) {
+        return _vault;
+    }
+
+    /// === Private and Internal ===
+    function _getPoolFromId(bytes32 poolId) internal pure returns (IManagedPool) {
+        // 12 byte logical shift left to remove the nonce and specialization setting. We don't need to mask,
+        // since the logical shift already sets the upper bits to zero.
+        return IManagedPool(address(uint256(poolId) >> (12 * 8)));
+    }
+
+    function _getPool() internal view returns (IManagedPool) {
+        return _getPoolFromId(getPoolId());
+    }
+
+    function _setTokens(IERC20[] memory tokens) internal {
+        _tokens = ComposablePoolLib.dropBptFromTokens(tokens);
+    }
+
+    // Returns the time until weights are updated
```

Comment:
Inaccurate comment -- up to you if you want to remove the comment or update the fn